File: src/tuner/simulator/xcp_simulator.py
from __future__ import annotations

import socket
import threading
from dataclasses import dataclass, field
import struct

from tuner.comms.xcp.packets import XcpCommand, XcpPid


@dataclass(slots=True)
class XcpSimulatorState:
    session_status: int = 0x05
    protection_status: int = 0x00
    configuration_status: int = 0x0001
    resource: int = 0x01
    comm_mode_basic: int = 0x02
    max_cto: int = 0x08
    max_dto: int = 0x0100
    protocol_layer_version: int = 0x01
    transport_layer_version: int = 0x01
    identifier: bytes = b"TUNERPY-XCP-SIM"
    memory: bytearray = field(default_factory=lambda: bytearray(256))

    def __post_init__(self) -> None:
        self.memory[0:4] = b"\x12\x34\x56\x78"
        self.memory[4:8] = (3210).to_bytes(4, byteorder="big", signed=False)
        self.memory[8:10] = (875).to_bytes(2, byteorder="big", signed=False)
        self.memory[10:14] = struct.pack(">f", 14.7)


class XcpSimulatorServer:
    def __init__(self, host: str = "127.0.0.1", port: int = 0) -> None:
        self.host = host
        self.port = port
        self.state = XcpSimulatorState()
        self._server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._thread: threading.Thread | None = None
        self._stop = threading.Event()
        self._mta_address = 0

    @property
    def address(self) -> tuple[str, int]:
        return self._server.getsockname()

    def start(self) -> None:
        self._server.bind((self.host, self.port))
        self._server.listen(1)
        self._thread = threading.Thread(target=self._serve, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        try:
            with socket.create_connection(self.address, timeout=0.2):
                pass
        except OSError:
            pass
        if self._thread is not None:
            self._thread.join(timeout=2.0)
        self._server.close()

    def _serve(self) -> None:
        while not self._stop.is_set():
            try:
                conn, _addr = self._server.accept()
            except OSError:
                return
            with conn:
                while not self._stop.is_set():
                    try:
                        command = self._read_command(conn)
                    except ConnectionResetError:
                        break
                    if not command:
                        break
                    response = self._handle(command)
                    try:
                        conn.sendall(response)
                    except ConnectionResetError:
                        break

    def _read_command(self, conn: socket.socket) -> bytes:
        header = conn.recv(1)
        if not header:
            return b""
        opcode = header[0]
        expected_size = self._expected_command_size(opcode)
        if expected_size == 1:
            return header
        payload = bytearray()
        while len(payload) < expected_size - 1:
            chunk = conn.recv(expected_size - 1 - len(payload))
            if not chunk:
                break
            payload.extend(chunk)
        return header + bytes(payload)

    @staticmethod
    def _expected_command_size(opcode: int) -> int:
        if opcode == XcpCommand.GET_STATUS:
            return 1
        if opcode in {XcpCommand.CONNECT, XcpCommand.GET_ID, XcpCommand.UPLOAD}:
            return 2
        if opcode == XcpCommand.SET_MTA:
            return 8
        return 1

    def _handle(self, packet: bytes) -> bytes:
        command = packet[0]
        if command == XcpCommand.CONNECT:
            return bytes(
                [
                    XcpPid.POSITIVE_RESPONSE,
                    self.state.resource,
                    self.state.comm_mode_basic,
                    self.state.max_cto,
                    (self.state.max_dto >> 8) & 0xFF,
                    self.state.max_dto & 0xFF,
                    self.state.protocol_layer_version,
                    self.state.transport_layer_version,
                ]
            )
        if command == XcpCommand.GET_STATUS:
            return bytes(
                [
                    XcpPid.POSITIVE_RESPONSE,
                    self.state.session_status,
                    self.state.protection_status,
                    (self.state.configuration_status >> 8) & 0xFF,
                    self.state.configuration_status & 0xFF,
                    0x00,
                ]
            )
        if command == XcpCommand.GET_ID:
            length = len(self.state.identifier)
            return (
                bytes(
                    [
                        XcpPid.POSITIVE_RESPONSE,
                        0x01,
                        0x00,
                        0x00,
                        (length >> 24) & 0xFF,
                        (length >> 16) & 0xFF,
                        (length >> 8) & 0xFF,
                        length & 0xFF,
                    ]
                )
                + self.state.identifier
            )
        if command == XcpCommand.SET_MTA:
            if len(packet) < 8:
                return bytes([0xFE, 0x20])
            self._mta_address = int.from_bytes(packet[4:8], byteorder="big", signed=False)
            return bytes([XcpPid.POSITIVE_RESPONSE])
        if command == XcpCommand.UPLOAD:
            if len(packet) < 2:
                return bytes([0xFE, 0x20])
            size = packet[1]
            data = bytes(self.state.memory[self._mta_address : self._mta_address + size])
            if len(data) < size:
                data = data + (b"\x00" * (size - len(data)))
            self._mta_address += size
            return bytes([XcpPid.POSITIVE_RESPONSE]) + data
        return bytes([0xFE, 0x20])

File: cpp/src/xcp_simulator.cpp
// SPDX-License-Identifier: MIT
//
// Implementation of `xcp_simulator.hpp`. Direct port of the
// command-dispatch helpers in `XcpSimulatorServer`.

#include "tuner_core/xcp_simulator.hpp"

#include "tuner_core/xcp_packets.hpp"

#include <cstring>

namespace tuner_core::xcp_simulator {

namespace {

// Pack 32-bit `value` big-endian into the 4-byte buffer at `dest`.
void store_be_u32(std::uint8_t* dest, std::uint32_t value) {
    dest[0] = static_cast<std::uint8_t>((value >> 24) & 0xFF);
    dest[1] = static_cast<std::uint8_t>((value >> 16) & 0xFF);
    dest[2] = static_cast<std::uint8_t>((value >> 8) & 0xFF);
    dest[3] = static_cast<std::uint8_t>(value & 0xFF);
}

void store_be_u16(std::uint8_t* dest, std::uint16_t value) {
    dest[0] = static_cast<std::uint8_t>((value >> 8) & 0xFF);
    dest[1] = static_cast<std::uint8_t>(value & 0xFF);
}

std::uint32_t load_be_u32(std::span<const std::uint8_t> bytes) {
    return (static_cast<std::uint32_t>(bytes[0]) << 24) |
           (static_cast<std::uint32_t>(bytes[1]) << 16) |
           (static_cast<std::uint32_t>(bytes[2]) <<  8) |
           (static_cast<std::uint32_t>(bytes[3]));
}

const std::vector<std::uint8_t> kError = {0xFE, 0x20};

}  // namespace

XcpSimulatorState XcpSimulatorState::default_state() {
    XcpSimulatorState s;
    s.memory.resize(256, 0);
    // memory[0:4] = 0x12 0x34 0x56 0x78
    s.memory[0] = 0x12;
    s.memory[1] = 0x34;
    s.memory[2] = 0x56;
    s.memory[3] = 0x78;
    // memory[4:8] = u32 big-endian 3210
    store_be_u32(s.memory.data() + 4, 3210u);
    // memory[8:10] = u16 big-endian 875
    store_be_u16(s.memory.data() + 8, 875u);
    // memory[10:14] = f32 big-endian 14.7
    {
        const float f = 14.7f;
        std::uint32_t bits = 0;
        std::memcpy(&bits, &f, sizeof(bits));
        store_be_u32(s.memory.data() + 10, bits);
    }
    return s;
}

int expected_command_size(std::uint8_t opcode) noexcept {
    using xcp_packets::XcpCommand;
    if (opcode == XcpCommand::GET_STATUS) return 1;
    if (opcode == XcpCommand::CONNECT ||
        opcode == XcpCommand::GET_ID  ||
        opcode == XcpCommand::UPLOAD) return 2;
    if (opcode == XcpCommand::SET_MTA) return 8;
    return 1;
}

DispatchResult handle_command(const XcpSimulatorState& state,
                              std::span<const std::uint8_t> packet,
                              std::uint32_t mta_address) {
    using xcp_packets::XcpCommand;
    using xcp_packets::XcpPid;

    DispatchResult result;
    result.new_mta_address = mta_address;

    if (packet.empty()) {
        result.response = kError;
        return result;
    }

    const std::uint8_t command = packet[0];

    if (command == XcpCommand::CONNECT) {
        result.response = {
            XcpPid::POSITIVE_RESPONSE,
            state.resource,
            state.comm_mode_basic,
            state.max_cto,
            static_cast<std::uint8_t>((state.max_dto >> 8) & 0xFF),
            static_cast<std::uint8_t>(state.max_dto & 0xFF),
            state.protocol_layer_version,
            state.transport_layer_version,
        };
        return result;
    }

    if (command == XcpCommand::GET_STATUS) {
        result.response = {
            XcpPid::POSITIVE_RESPONSE,
            state.session_status,
            state.protection_status,
            static_cast<std::uint8_t>((state.configuration_status >> 8) & 0xFF),
            static_cast<std::uint8_t>(state.configuration_status & 0xFF),
            0x00,
        };
        return result;
    }

    if (command == XcpCommand::GET_ID) {
        const std::uint32_t length = static_cast<std::uint32_t>(state.identifier.size());
        result.response = {
            XcpPid::POSITIVE_RESPONSE,
            0x01,
            0x00,
            0x00,
            static_cast<std::uint8_t>((length >> 24) & 0xFF),
            static_cast<std::uint8_t>((length >> 16) & 0xFF),
            static_cast<std::uint8_t>((length >>  8) & 0xFF),
            static_cast<std::uint8_t>(length & 0xFF),
        };
        result.response.insert(result.response.end(),
                               state.identifier.begin(),
                               state.identifier.end());
        return result;
    }

    if (command == XcpCommand::SET_MTA) {
        if (packet.size() < 8) {
            result.response = kError;
            return result;
        }
        result.new_mta_address = load_be_u32(packet.subspan(4, 4));
        result.response = {XcpPid::POSITIVE_RESPONSE};
        return result;
    }

    if (command == XcpCommand::UPLOAD) {
        if (packet.size() < 2) {
            result.response = kError;
            return result;
        }
        const std::uint8_t size = packet[1];
        result.response.reserve(static_cast<std::size_t>(size) + 1);
        result.response.push_back(XcpPid::POSITIVE_RESPONSE);
        // Read up to `size` bytes from memory[mta..mta+size); zero-pad
        // anything past the end of the memory buffer (mirrors the
        // Python `data + b"\x00" * (size - len(data))`).
        const std::uint32_t mem_size = static_cast<std::uint32_t>(state.memory.size());
        for (std::uint32_t i = 0; i < size; ++i) {
            const std::uint32_t addr = mta_address + i;
            if (addr < mem_size) {
                result.response.push_back(state.memory[addr]);
            } else {
                result.response.push_back(0x00);
            }
        }
        result.new_mta_address = mta_address + size;
        return result;
    }

    result.response = kError;
    return result;
}

}  // namespace tuner_core::xcp_simulator
